driver/format/RowBinaryWithNamesAndTypes.cpp

```diff
@@ -0,0 +1,346 @@
+        case DataSourceTypeId::Decimal64:   return readValueAs<DataSourceType< DataSourceTypeId::Decimal64   >>(dest, column_info);
+        case DataSourceTypeId::Decimal128:  return readValueAs<DataSourceType< DataSourceTypeId::Decimal128  >>(dest, column_info);
+        case DataSourceTypeId::FixedString: return readValueAs<DataSourceType< DataSourceTypeId::FixedString >>(dest, column_info);
+        case DataSourceTypeId::Float32:     return readValueAs<DataSourceType< DataSourceTypeId::Float32     >>(dest, column_info);
+        case DataSourceTypeId::Float64:     return readValueAs<DataSourceType< DataSourceTypeId::Float64     >>(dest, column_info);
+        case DataSourceTypeId::Int8:        return readValueAs<DataSourceType< DataSourceTypeId::Int8        >>(dest, column_info);
+        case DataSourceTypeId::Int16:       return readValueAs<DataSourceType< DataSourceTypeId::Int16       >>(dest, column_info);
+        case DataSourceTypeId::Int32:       return readValueAs<DataSourceType< DataSourceTypeId::Int32       >>(dest, column_info);
+        case DataSourceTypeId::Int64:       return readValueAs<DataSourceType< DataSourceTypeId::Int64       >>(dest, column_info);
+        case DataSourceTypeId::Nothing:     return readValueAs<DataSourceType< DataSourceTypeId::Nothing     >>(dest, column_info);
+        case DataSourceTypeId::String:      return readValueAs<DataSourceType< DataSourceTypeId::String      >>(dest, column_info);
+        case DataSourceTypeId::UInt8:       return readValueAs<DataSourceType< DataSourceTypeId::UInt8       >>(dest, column_info);
+        case DataSourceTypeId::UInt16:      return readValueAs<DataSourceType< DataSourceTypeId::UInt16      >>(dest, column_info);
+        case DataSourceTypeId::UInt32:      return readValueAs<DataSourceType< DataSourceTypeId::UInt32      >>(dest, column_info);
+        case DataSourceTypeId::UInt64:      return readValueAs<DataSourceType< DataSourceTypeId::UInt64      >>(dest, column_info);
+        case DataSourceTypeId::UUID:        return readValueAs<DataSourceType< DataSourceTypeId::UUID        >>(dest, column_info);
+        default:                            throw std::runtime_error("Unable to decode value of type '" + column_info.type + "'");
+    }
+}
+
+void RowBinaryWithNamesAndTypesResultSet::readValue(DataSourceType<DataSourceTypeId::Date> & dest, ColumnInfo & column_info) {
+    std::uint16_t days_since_epoch = 0;
+    readPOD(days_since_epoch);
+
+    std::time_t time = days_since_epoch;
+    time = time * 24 * 60 * 60; // Now it's seconds since epoch.
+    const auto & tm = *std::localtime(&time);
+
+    dest.value.year = 1900 + tm.tm_year;
+    dest.value.month = 1 + tm.tm_mon;
+    dest.value.day = tm.tm_mday;
+}
+
+void RowBinaryWithNamesAndTypesResultSet::readValue(DataSourceType<DataSourceTypeId::DateTime> & dest, ColumnInfo & column_info) {
```

Comment:
Please validate that Date and DateTime values returned by the ODBC driver are equal to the value returned by ClickHouse client when local timezone differs from timezone of server.